{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f35eb7ee-6dc4-4256-ae28-fba103e486bb",
   "metadata": {},
   "source": [
    "## Load and process negative GTeX SNPs from the eQTL catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "77452b44-7e8f-40bb-8a28-26896e8a4c4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import anndata\n",
    "import os\n",
    "import tqdm\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac90082a-dc89-4df2-8197-77dfba377b4c",
   "metadata": {},
   "source": [
    "## Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "934cc48e-97bf-4275-9789-41fa0fe59b80",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_cat_path = 'https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/master/data_tables/dataset_metadata.tsv'\n",
    "anndata_file = \"/gstore/data/resbioai/grelu/decima/20240823/data.h5ad\"\n",
    "susie_file='/gstore/data/resbioai/grelu/decima/20240823/bulk_eqtl_results/susie_df.csv'\n",
    "cs_vars_file='/gstore/data/resbioai/grelu/decima/20240823/bulk_eqtl_results/cs_vars.npy'\n",
    "out_dir = '/gstore/data/resbioai/grelu/decima/20240823/bulk_eqtl_results/'\n",
    "sumstats_dir ='/gstore/data/resbioai/grelu/decima/QTS000015/sumstats/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "648e2372-5e6f-4232-b997-b713538756a2",
   "metadata": {},
   "source": [
    "## Get GTeX dataset ID to tissue mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b555d8ff-1fc9-496e-bbd7-b11b2a608726",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>sample_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>QTD000121</td>\n",
       "      <td>adipose_visceral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>QTD000126</td>\n",
       "      <td>adrenal_gland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>QTD000131</td>\n",
       "      <td>artery_aorta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>QTD000136</td>\n",
       "      <td>artery_coronary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>QTD000141</td>\n",
       "      <td>artery_tibial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>QTD000146</td>\n",
       "      <td>brain_amygdala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>QTD000151</td>\n",
       "      <td>brain_anterior_cingulate_cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>QTD000156</td>\n",
       "      <td>brain_caudate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>QTD000161</td>\n",
       "      <td>brain_cerebellar_hemisphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>QTD000166</td>\n",
       "      <td>brain_cerebellum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>QTD000171</td>\n",
       "      <td>brain_cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>QTD000176</td>\n",
       "      <td>brain_frontal_cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>QTD000181</td>\n",
       "      <td>brain_hippocampus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>QTD000186</td>\n",
       "      <td>brain_hypothalamus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>QTD000191</td>\n",
       "      <td>brain_nucleus_accumbens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>QTD000196</td>\n",
       "      <td>brain_putamen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>QTD000201</td>\n",
       "      <td>brain_spinal_cord</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>QTD000206</td>\n",
       "      <td>brain_substantia_nigra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>QTD000211</td>\n",
       "      <td>breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>QTD000216</td>\n",
       "      <td>fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>QTD000221</td>\n",
       "      <td>LCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>QTD000226</td>\n",
       "      <td>colon_sigmoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>QTD000231</td>\n",
       "      <td>colon_transverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>QTD000236</td>\n",
       "      <td>esophagus_gej</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>QTD000241</td>\n",
       "      <td>esophagus_mucosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>QTD000246</td>\n",
       "      <td>esophagus_muscularis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>QTD000251</td>\n",
       "      <td>heart_atrial_appendage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>QTD000256</td>\n",
       "      <td>heart_left_ventricle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>QTD000261</td>\n",
       "      <td>kidney_cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>QTD000266</td>\n",
       "      <td>liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>QTD000271</td>\n",
       "      <td>lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>QTD000276</td>\n",
       "      <td>minor_salivary_gland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>QTD000281</td>\n",
       "      <td>muscle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>QTD000286</td>\n",
       "      <td>nerve_tibial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>QTD000291</td>\n",
       "      <td>ovary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>QTD000296</td>\n",
       "      <td>pancreas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>QTD000301</td>\n",
       "      <td>pituitary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>QTD000306</td>\n",
       "      <td>prostate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>QTD000311</td>\n",
       "      <td>skin_not_sun_exposed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>QTD000316</td>\n",
       "      <td>skin_sun_exposed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>QTD000321</td>\n",
       "      <td>small_intestine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>QTD000326</td>\n",
       "      <td>spleen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>QTD000331</td>\n",
       "      <td>stomach</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>QTD000336</td>\n",
       "      <td>testis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>QTD000341</td>\n",
       "      <td>thyroid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>QTD000346</td>\n",
       "      <td>uterus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>QTD000351</td>\n",
       "      <td>vagina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>QTD000356</td>\n",
       "      <td>blood</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    dataset_id                     sample_group\n",
       "115  QTD000116             adipose_subcutaneous\n",
       "120  QTD000121                 adipose_visceral\n",
       "125  QTD000126                    adrenal_gland\n",
       "130  QTD000131                     artery_aorta\n",
       "135  QTD000136                  artery_coronary\n",
       "140  QTD000141                    artery_tibial\n",
       "145  QTD000146                   brain_amygdala\n",
       "150  QTD000151  brain_anterior_cingulate_cortex\n",
       "155  QTD000156                    brain_caudate\n",
       "160  QTD000161      brain_cerebellar_hemisphere\n",
       "165  QTD000166                 brain_cerebellum\n",
       "170  QTD000171                     brain_cortex\n",
       "175  QTD000176             brain_frontal_cortex\n",
       "180  QTD000181                brain_hippocampus\n",
       "185  QTD000186               brain_hypothalamus\n",
       "190  QTD000191          brain_nucleus_accumbens\n",
       "195  QTD000196                    brain_putamen\n",
       "200  QTD000201                brain_spinal_cord\n",
       "205  QTD000206           brain_substantia_nigra\n",
       "210  QTD000211                           breast\n",
       "215  QTD000216                       fibroblast\n",
       "220  QTD000221                              LCL\n",
       "225  QTD000226                    colon_sigmoid\n",
       "230  QTD000231                 colon_transverse\n",
       "235  QTD000236                    esophagus_gej\n",
       "240  QTD000241                 esophagus_mucosa\n",
       "245  QTD000246             esophagus_muscularis\n",
       "250  QTD000251           heart_atrial_appendage\n",
       "255  QTD000256             heart_left_ventricle\n",
       "260  QTD000261                    kidney_cortex\n",
       "265  QTD000266                            liver\n",
       "270  QTD000271                             lung\n",
       "275  QTD000276             minor_salivary_gland\n",
       "280  QTD000281                           muscle\n",
       "285  QTD000286                     nerve_tibial\n",
       "290  QTD000291                            ovary\n",
       "295  QTD000296                         pancreas\n",
       "300  QTD000301                        pituitary\n",
       "305  QTD000306                         prostate\n",
       "310  QTD000311             skin_not_sun_exposed\n",
       "315  QTD000316                 skin_sun_exposed\n",
       "320  QTD000321                  small_intestine\n",
       "325  QTD000326                           spleen\n",
       "330  QTD000331                          stomach\n",
       "335  QTD000336                           testis\n",
       "340  QTD000341                          thyroid\n",
       "345  QTD000346                           uterus\n",
       "350  QTD000351                           vagina\n",
       "355  QTD000356                            blood"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_meta = pd.read_table(eqtl_cat_path)\n",
    "eqtl_meta = eqtl_meta[eqtl_meta.quant_method == 'ge'] # gene exp. QTLs\n",
    "eqtl_meta = eqtl_meta[eqtl_meta.study_label == 'GTEx']\n",
    "eqtl_meta = eqtl_meta[['dataset_id', 'sample_group']]\n",
    "\n",
    "eqtl_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4eb24b79-e097-40c7-8709-592897c7ca06",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'QTD000116': 'adipose_subcutaneous',\n",
       " 'QTD000121': 'adipose_visceral',\n",
       " 'QTD000126': 'adrenal_gland',\n",
       " 'QTD000131': 'artery_aorta',\n",
       " 'QTD000136': 'artery_coronary',\n",
       " 'QTD000141': 'artery_tibial',\n",
       " 'QTD000146': 'brain_amygdala',\n",
       " 'QTD000151': 'brain_anterior_cingulate_cortex',\n",
       " 'QTD000156': 'brain_caudate',\n",
       " 'QTD000161': 'brain_cerebellar_hemisphere',\n",
       " 'QTD000166': 'brain_cerebellum',\n",
       " 'QTD000171': 'brain_cortex',\n",
       " 'QTD000176': 'brain_frontal_cortex',\n",
       " 'QTD000181': 'brain_hippocampus',\n",
       " 'QTD000186': 'brain_hypothalamus',\n",
       " 'QTD000191': 'brain_nucleus_accumbens',\n",
       " 'QTD000196': 'brain_putamen',\n",
       " 'QTD000201': 'brain_spinal_cord',\n",
       " 'QTD000206': 'brain_substantia_nigra',\n",
       " 'QTD000211': 'breast',\n",
       " 'QTD000216': 'fibroblast',\n",
       " 'QTD000221': 'LCL',\n",
       " 'QTD000226': 'colon_sigmoid',\n",
       " 'QTD000231': 'colon_transverse',\n",
       " 'QTD000236': 'esophagus_gej',\n",
       " 'QTD000241': 'esophagus_mucosa',\n",
       " 'QTD000246': 'esophagus_muscularis',\n",
       " 'QTD000251': 'heart_atrial_appendage',\n",
       " 'QTD000256': 'heart_left_ventricle',\n",
       " 'QTD000261': 'kidney_cortex',\n",
       " 'QTD000266': 'liver',\n",
       " 'QTD000271': 'lung',\n",
       " 'QTD000276': 'minor_salivary_gland',\n",
       " 'QTD000281': 'muscle',\n",
       " 'QTD000286': 'nerve_tibial',\n",
       " 'QTD000291': 'ovary',\n",
       " 'QTD000296': 'pancreas',\n",
       " 'QTD000301': 'pituitary',\n",
       " 'QTD000306': 'prostate',\n",
       " 'QTD000311': 'skin_not_sun_exposed',\n",
       " 'QTD000316': 'skin_sun_exposed',\n",
       " 'QTD000321': 'small_intestine',\n",
       " 'QTD000326': 'spleen',\n",
       " 'QTD000331': 'stomach',\n",
       " 'QTD000336': 'testis',\n",
       " 'QTD000341': 'thyroid',\n",
       " 'QTD000346': 'uterus',\n",
       " 'QTD000351': 'vagina',\n",
       " 'QTD000356': 'blood'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct_dict = eqtl_meta.set_index('dataset_id').sample_group.to_dict()\n",
    "ct_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcaffcdc-c144-4b35-b0bb-3c9e979f60a4",
   "metadata": {},
   "source": [
    "## Load decima metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "faab12c8-0db9-4894-960e-d0f52e22df57",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad = anndata.read_h5ad(anndata_file)\n",
    "ensembl_id_map = ad.var[['gene_id']].reset_index().set_index('gene_id')['index'].to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f08549e-cbd6-45ba-9cca-4898f2ae95c0",
   "metadata": {},
   "source": [
    "## Load negative variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c48ea7d8-9fa1-4b00-a787-5fe1fdb2a5c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each gene, we need the interval, to exclude \"unseeable\" stuff\n",
    "gene_to_interval_dict = {v.gene_id:{'chr':v.chrom,'start':v.start,'end':v.end} for k,v in ad.var.iterrows()}\n",
    "\n",
    "def check_scoreability(row):\n",
    "    if row['gene_id'] not in gene_to_interval_dict:\n",
    "        return False\n",
    "    interval = gene_to_interval_dict[row['gene_id']]\n",
    "    pos = row['position']\n",
    "    gene_window_start = interval['start']\n",
    "    gene_window_end = interval['end']\n",
    "    return (pos > gene_window_start) & (pos < gene_window_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ad808882-0286-4ca5-b482-31262fc5cd31",
   "metadata": {},
   "outputs": [],
   "source": [
    "susie_df = pd.read_csv(susie_file)\n",
    "cs_vars = np.load(cs_vars_file)\n",
    "len(cs_vars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "97d02006-da67-49c3-889a-62a70808f1af",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QTD000116\n",
      "Loading sumstats\n",
      "168441427\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "455807\n",
      "QTD000121\n",
      "Loading sumstats\n",
      "167019167\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "298493\n",
      "QTD000126\n",
      "Loading sumstats\n",
      "155295380\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "193833\n",
      "QTD000131\n",
      "Loading sumstats\n",
      "156857426\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "302831\n",
      "QTD000136\n",
      "Loading sumstats\n",
      "163011525\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "115800\n",
      "QTD000141\n",
      "Loading sumstats\n",
      "156641820\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "444148\n",
      "QTD000146\n",
      "Loading sumstats\n",
      "142903869\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "59664\n",
      "QTD000151\n",
      "Loading sumstats\n",
      "144577410\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "93223\n",
      "QTD000156\n",
      "Loading sumstats\n",
      "170781268\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "162185\n",
      "QTD000161\n",
      "Loading sumstats\n",
      "146116003\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "256595\n",
      "QTD000166\n",
      "Loading sumstats\n",
      "157845385\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "352918\n",
      "QTD000171\n",
      "Loading sumstats\n",
      "163017375\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "176158\n",
      "QTD000176\n",
      "Loading sumstats\n",
      "152628125\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "128443\n",
      "QTD000181\n",
      "Loading sumstats\n",
      "157111948\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "76039\n",
      "QTD000186\n",
      "Loading sumstats\n",
      "162109365\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "94021\n",
      "QTD000191\n",
      "Loading sumstats\n",
      "162828702\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "140524\n",
      "QTD000196\n",
      "Loading sumstats\n",
      "159051325\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "137139\n",
      "QTD000201\n",
      "Loading sumstats\n",
      "145204842\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "109342\n",
      "QTD000206\n",
      "Loading sumstats\n",
      "138787026\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "59443\n",
      "QTD000211\n",
      "Loading sumstats\n",
      "178536669\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "231309\n",
      "QTD000216\n",
      "Loading sumstats\n",
      "141088575\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "497598\n",
      "QTD000221\n",
      "Loading sumstats\n",
      "123444186\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "108353\n",
      "QTD000226\n",
      "Loading sumstats\n",
      "164902640\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "215528\n",
      "QTD000231\n",
      "Loading sumstats\n",
      "174219786\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "262885\n",
      "QTD000236\n",
      "Loading sumstats\n",
      "163443969\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "258348\n",
      "QTD000241\n",
      "Loading sumstats\n",
      "156605909\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "377523\n",
      "QTD000246\n",
      "Loading sumstats\n",
      "162128605\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "422182\n",
      "QTD000251\n",
      "Loading sumstats\n",
      "160828411\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "281644\n",
      "QTD000256\n",
      "Loading sumstats\n",
      "158138048\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "257032\n",
      "QTD000261\n",
      "Loading sumstats\n",
      "129380282\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "20556\n",
      "QTD000266\n",
      "Loading sumstats\n",
      "135974843\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "105113\n",
      "QTD000271\n",
      "Loading sumstats\n",
      "174485591\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "331437\n",
      "QTD000276\n",
      "Loading sumstats\n",
      "143294970\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "105038\n",
      "QTD000281\n",
      "Loading sumstats\n",
      "139517466\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "386540\n",
      "QTD000286\n",
      "Loading sumstats\n",
      "175982593\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "616474\n",
      "QTD000291\n",
      "Loading sumstats\n",
      "156059575\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "114127\n",
      "QTD000296\n",
      "Loading sumstats\n",
      "135889132\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "296178\n",
      "QTD000301\n",
      "Loading sumstats\n",
      "163548551\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "176336\n",
      "QTD000306\n",
      "Loading sumstats\n",
      "181343664\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "131825\n",
      "QTD000311\n",
      "Loading sumstats\n",
      "169056504\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "359722\n",
      "QTD000316\n",
      "Loading sumstats\n",
      "168248503\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "436253\n",
      "QTD000321\n",
      "Loading sumstats\n",
      "161886158\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "116137\n",
      "QTD000326\n",
      "Loading sumstats\n",
      "161166640\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "270290\n",
      "QTD000331\n",
      "Loading sumstats\n",
      "169420757\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "195413\n",
      "QTD000336\n",
      "Loading sumstats\n",
      "223643551\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "425272\n",
      "QTD000341\n",
      "Loading sumstats\n",
      "176857905\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "589887\n",
      "QTD000346\n",
      "Loading sumstats\n",
      "140173857\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "78993\n",
      "QTD000351\n",
      "Loading sumstats\n",
      "150895092\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "54872\n",
      "QTD000356\n",
      "Loading sumstats\n",
      "123192057\n",
      "Filtering negatives\n",
      "Filtering chromosomes\n",
      "Adding metadata\n",
      "411795\n"
     ]
    }
   ],
   "source": [
    "negvar_list = []\n",
    "\n",
    "for ds in ct_dict.keys():\n",
    "    print(ds)\n",
    "\n",
    "    print(\"Loading sumstats\")\n",
    "    file = f'{sumstats_dir}/{ds}/{ds}.all.tsv.gz'\n",
    "    negvars = pl.read_csv(file, separator=\"\\t\", \n",
    "                      schema_overrides={'chromosome':pl.String},\n",
    "                      columns=(1,2,3,4,5,7,8,11,16,18),\n",
    "                     )\n",
    "    print(len(negvars))\n",
    "\n",
    "    print(\"Filtering negatives\")\n",
    "    # Select common SNPs\n",
    "    negvars = negvars.filter(\n",
    "        (pl.col('pvalue') > 0.05) & (pl.col('type') == \"SNP\") & (pl.col('maf') > 0.05)) # pre-select nonsignificant\n",
    "    # For the same genes as the positives\n",
    "    negvars = negvars.filter(pl.col('gene_id').is_in(susie_df[susie_df.dataset == ds].gene_id))\n",
    "    # Not in any credible set\n",
    "    negvars = negvars.filter(~pl.col('variant').is_in(list(cs_vars)))\n",
    "    # In decima interval\n",
    "    negvars = negvars.filter(pl.struct(pl.all()).map_elements(check_scoreability, return_dtype=pl.Boolean))\n",
    "    \n",
    "    negvars = negvars.to_pandas()\n",
    "\n",
    "    print(\"Filtering chromosomes\")\n",
    "    negvars = negvars.rename(columns={'chromosome':'chrom', 'position':'pos'})\n",
    "    negvars = negvars.dropna(subset=['chrom'])\n",
    "    negvars['chrom'] = ['chr' + x for x in negvars.chrom]\n",
    "\n",
    "    print(\"Adding metadata\")\n",
    "    negvars['dataset'] = ds\n",
    "    negvars['celltype'] = ct_dict[ds]\n",
    "\n",
    "    print(len(negvars))\n",
    "    negvar_list.append(negvars)\n",
    "\n",
    "negvars = pd.concat(negvar_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5fbdbb6-8bb3-461b-87f1-058f8574dbf5",
   "metadata": {},
   "source": [
    "## Add gene metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cd660b46-2a6a-47c8-8fa2-bd52596789f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each negative, compute the relpos and distance to the TSS \n",
    "negvar_dedup = negvars[['gene_id','variant','pos']].drop_duplicates()\n",
    "\n",
    "# add gene information and calculate relative variant positions (offset)\n",
    "negvar_dedup = negvar_dedup.merge(ad.var[['gene_id', 'start', 'end', 'strand', 'gene_mask_start']]).rename(\n",
    "    columns={'start': 'gene_window_start', 'end': 'gene_window_end', 'strand': 'gene_strand'}) # add window information\n",
    "negvar_dedup['pos_relative'] = negvar_dedup.pos - negvar_dedup.gene_window_start - 1\n",
    "\n",
    "# use gene_end to calculate offset for - genes \n",
    "negvar_dedup.loc[negvar_dedup.gene_strand=='-', 'pos_relative'] = negvar_dedup.gene_window_end[negvar_dedup.gene_strand=='-'] - negvar_dedup.pos[negvar_dedup.gene_strand=='-']\n",
    "negvar_dedup['abspos_rel_TSS'] = np.abs(negvar_dedup[\"pos_relative\"] - negvar_dedup['gene_mask_start'])\n",
    "\n",
    "negvars = negvars.merge(negvar_dedup[['gene_id','variant','gene_strand',\"pos_relative\",\"abspos_rel_TSS\"]],\n",
    "    on=['gene_id','variant'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1742c4e2-9d8b-4774-a779-027d55d18877",
   "metadata": {},
   "source": [
    "## Add gene name from Decima metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "47a39ff8-3300-4528-ad0e-5a84496a602b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>dataset</th>\n",
       "      <th>celltype</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>pos_relative</th>\n",
       "      <th>abspos_rel_TSS</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>502653</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_502653_G_T</td>\n",
       "      <td>0.065405</td>\n",
       "      <td>0.586796</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs1167381111</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>488709</td>\n",
       "      <td>324869</td>\n",
       "      <td>LINC00115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>591460</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_591460_T_C</td>\n",
       "      <td>0.060241</td>\n",
       "      <td>0.831717</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs60396226</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>399902</td>\n",
       "      <td>236062</td>\n",
       "      <td>LINC00115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>625776</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1_625776_T_G</td>\n",
       "      <td>0.129088</td>\n",
       "      <td>0.398739</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs150334593</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>365586</td>\n",
       "      <td>201746</td>\n",
       "      <td>LINC00115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>627983</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_627983_C_T</td>\n",
       "      <td>0.133391</td>\n",
       "      <td>0.339284</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs137905425</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>363379</td>\n",
       "      <td>199539</td>\n",
       "      <td>LINC00115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>632781</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_632781_C_T</td>\n",
       "      <td>0.149742</td>\n",
       "      <td>0.195945</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs61769296</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>358581</td>\n",
       "      <td>194741</td>\n",
       "      <td>LINC00115</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     pos ref alt          variant       maf    pvalue          gene_id  \\\n",
       "0  chr1  502653   G   T  chr1_502653_G_T  0.065405  0.586796  ENSG00000225880   \n",
       "1  chr1  591460   T   C  chr1_591460_T_C  0.060241  0.831717  ENSG00000225880   \n",
       "2  chr1  625776   T   G  chr1_625776_T_G  0.129088  0.398739  ENSG00000225880   \n",
       "3  chr1  627983   C   T  chr1_627983_C_T  0.133391  0.339284  ENSG00000225880   \n",
       "4  chr1  632781   C   T  chr1_632781_C_T  0.149742  0.195945  ENSG00000225880   \n",
       "\n",
       "           rsid    dataset              celltype gene_strand  pos_relative  \\\n",
       "0  rs1167381111  QTD000116  adipose_subcutaneous           -        488709   \n",
       "1    rs60396226  QTD000116  adipose_subcutaneous           -        399902   \n",
       "2   rs150334593  QTD000116  adipose_subcutaneous           -        365586   \n",
       "3   rs137905425  QTD000116  adipose_subcutaneous           -        363379   \n",
       "4    rs61769296  QTD000116  adipose_subcutaneous           -        358581   \n",
       "\n",
       "   abspos_rel_TSS       gene  \n",
       "0          324869  LINC00115  \n",
       "1          236062  LINC00115  \n",
       "2          201746  LINC00115  \n",
       "3          199539  LINC00115  \n",
       "4          194741  LINC00115  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negvars = negvars.drop(columns=['type'])\n",
    "negvars['gene'] = negvars.gene_id.map(ensembl_id_map)\n",
    "negvars.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecc3ae73-6309-4124-98b9-aa3bef8a6252",
   "metadata": {},
   "source": [
    "## Add credible set columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4dbd9c01-53e7-4836-b25a-16018909ac50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>maf</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>dataset</th>\n",
       "      <th>celltype</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>pos_relative</th>\n",
       "      <th>abspos_rel_TSS</th>\n",
       "      <th>gene</th>\n",
       "      <th>pip</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>cs_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>502653</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_502653_G_T</td>\n",
       "      <td>0.065405</td>\n",
       "      <td>0.586796</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs1167381111</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>488709</td>\n",
       "      <td>324869</td>\n",
       "      <td>LINC00115</td>\n",
       "      <td>0</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>591460</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_591460_T_C</td>\n",
       "      <td>0.060241</td>\n",
       "      <td>0.831717</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs60396226</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>399902</td>\n",
       "      <td>236062</td>\n",
       "      <td>LINC00115</td>\n",
       "      <td>0</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>625776</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1_625776_T_G</td>\n",
       "      <td>0.129088</td>\n",
       "      <td>0.398739</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs150334593</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>365586</td>\n",
       "      <td>201746</td>\n",
       "      <td>LINC00115</td>\n",
       "      <td>0</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>627983</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_627983_C_T</td>\n",
       "      <td>0.133391</td>\n",
       "      <td>0.339284</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs137905425</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>363379</td>\n",
       "      <td>199539</td>\n",
       "      <td>LINC00115</td>\n",
       "      <td>0</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>632781</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_632781_C_T</td>\n",
       "      <td>0.149742</td>\n",
       "      <td>0.195945</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>rs61769296</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>-</td>\n",
       "      <td>358581</td>\n",
       "      <td>194741</td>\n",
       "      <td>LINC00115</td>\n",
       "      <td>0</td>\n",
       "      <td>negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     pos ref alt          variant       maf    pvalue          gene_id  \\\n",
       "0  chr1  502653   G   T  chr1_502653_G_T  0.065405  0.586796  ENSG00000225880   \n",
       "1  chr1  591460   T   C  chr1_591460_T_C  0.060241  0.831717  ENSG00000225880   \n",
       "2  chr1  625776   T   G  chr1_625776_T_G  0.129088  0.398739  ENSG00000225880   \n",
       "3  chr1  627983   C   T  chr1_627983_C_T  0.133391  0.339284  ENSG00000225880   \n",
       "4  chr1  632781   C   T  chr1_632781_C_T  0.149742  0.195945  ENSG00000225880   \n",
       "\n",
       "           rsid    dataset              celltype gene_strand  pos_relative  \\\n",
       "0  rs1167381111  QTD000116  adipose_subcutaneous           -        488709   \n",
       "1    rs60396226  QTD000116  adipose_subcutaneous           -        399902   \n",
       "2   rs150334593  QTD000116  adipose_subcutaneous           -        365586   \n",
       "3   rs137905425  QTD000116  adipose_subcutaneous           -        363379   \n",
       "4    rs61769296  QTD000116  adipose_subcutaneous           -        358581   \n",
       "\n",
       "   abspos_rel_TSS       gene  pip     cs_id  cs_size  \n",
       "0          324869  LINC00115    0  negative        0  \n",
       "1          236062  LINC00115    0  negative        0  \n",
       "2          201746  LINC00115    0  negative        0  \n",
       "3          199539  LINC00115    0  negative        0  \n",
       "4          194741  LINC00115    0  negative        0  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negvars['pip'] = 0\n",
    "negvars['cs_id'] = 'negative'\n",
    "negvars['cs_size'] = 0\n",
    "negvars.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa0117ba-b0ca-4157-9b09-40da5feb4492",
   "metadata": {},
   "source": [
    "## Match to positive variants by distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "61c20085-50df-48e0-a191-fde47d954734",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "brain_caudate\n",
      "adipose_subcutaneous\n",
      "heart_left_ventricle\n",
      "artery_tibial\n",
      "vagina\n",
      "brain_hypothalamus\n",
      "brain_nucleus_accumbens\n",
      "stomach\n",
      "pituitary\n",
      "pancreas\n",
      "breast\n",
      "uterus\n",
      "testis\n",
      "esophagus_mucosa\n",
      "esophagus_muscularis\n",
      "brain_amygdala\n",
      "artery_aorta\n",
      "brain_cerebellum\n",
      "colon_sigmoid\n",
      "kidney_cortex\n",
      "ovary\n",
      "fibroblast\n",
      "brain_frontal_cortex\n",
      "adipose_visceral\n",
      "heart_atrial_appendage\n",
      "nerve_tibial\n",
      "brain_cerebellar_hemisphere\n",
      "colon_transverse\n",
      "lung\n",
      "artery_coronary\n",
      "brain_cortex\n",
      "esophagus_gej\n",
      "brain_anterior_cingulate_cortex\n",
      "thyroid\n",
      "skin_sun_exposed\n",
      "brain_substantia_nigra\n",
      "blood\n",
      "small_intestine\n",
      "prostate\n",
      "skin_not_sun_exposed\n",
      "brain_putamen\n",
      "brain_spinal_cord\n",
      "adrenal_gland\n",
      "minor_salivary_gland\n",
      "spleen\n",
      "liver\n",
      "LCL\n",
      "brain_hippocampus\n",
      "muscle\n"
     ]
    }
   ],
   "source": [
    "# for each positive variant, collect target_negative_n negatives wich are as close to the TSS as possible, and not yet selected for this cell-type\n",
    "target_negative_n = 20\n",
    "selected_vars = {}\n",
    "for celltype in set(susie_df['celltype']):\n",
    "    print(celltype)\n",
    "    \n",
    "    curr_positive_df = susie_df.query('celltype == @celltype')\n",
    "    curr_negative_df = negvars.query('celltype == @celltype')\n",
    "    \n",
    "    selected_vars[celltype] = set()\n",
    "    for _,positive in curr_positive_df.iterrows():\n",
    "        negative_sub = curr_negative_df[curr_negative_df.gene_id == positive['gene_id']]\n",
    "        negative_sub = negative_sub.sort_values('abspos_rel_TSS')\n",
    "        i = 0\n",
    "        for _,variant in negative_sub.iterrows():\n",
    "            if variant['variant'] not in selected_vars[celltype]:\n",
    "                selected_vars[celltype].add(variant['variant'])\n",
    "                i += 1\n",
    "                if i == target_negative_n:\n",
    "                    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fd4c6e0f-9616-49fd-8f21-c72ed9a1586a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19049, 523679)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negvars_reduced = pd.concat([\n",
    "    negvars[negvars.variant.isin(selected_vars[celltype]) & (negvars.celltype == celltype)\n",
    "        ] for celltype in selected_vars])\n",
    "len(susie_df), len(negvars_reduced)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e57bdf1c-3d45-4449-9bbf-4b33b939975b",
   "metadata": {},
   "source": [
    "## Print summary stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9b2e7f62-e8ca-47a2-b896-73f9318431d9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>pos_genes</th>\n",
       "      <th>pos_genes_with_neg_ct</th>\n",
       "      <th>total_pos</th>\n",
       "      <th>total_matched_neg</th>\n",
       "      <th>reduced_matched_neg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>brain_caudate</td>\n",
       "      <td>237</td>\n",
       "      <td>237</td>\n",
       "      <td>245</td>\n",
       "      <td>145892</td>\n",
       "      <td>4926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>695</td>\n",
       "      <td>695</td>\n",
       "      <td>748</td>\n",
       "      <td>381142</td>\n",
       "      <td>15206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>heart_left_ventricle</td>\n",
       "      <td>364</td>\n",
       "      <td>364</td>\n",
       "      <td>364</td>\n",
       "      <td>225530</td>\n",
       "      <td>7620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>artery_tibial</td>\n",
       "      <td>684</td>\n",
       "      <td>684</td>\n",
       "      <td>729</td>\n",
       "      <td>371392</td>\n",
       "      <td>14887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>vagina</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>84</td>\n",
       "      <td>51042</td>\n",
       "      <td>1686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brain_hypothalamus</td>\n",
       "      <td>134</td>\n",
       "      <td>134</td>\n",
       "      <td>139</td>\n",
       "      <td>87201</td>\n",
       "      <td>2802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>brain_nucleus_accumbens</td>\n",
       "      <td>203</td>\n",
       "      <td>203</td>\n",
       "      <td>207</td>\n",
       "      <td>127664</td>\n",
       "      <td>4211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stomach</td>\n",
       "      <td>299</td>\n",
       "      <td>299</td>\n",
       "      <td>310</td>\n",
       "      <td>171579</td>\n",
       "      <td>6342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pituitary</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>264</td>\n",
       "      <td>161558</td>\n",
       "      <td>5374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pancreas</td>\n",
       "      <td>421</td>\n",
       "      <td>421</td>\n",
       "      <td>435</td>\n",
       "      <td>264308</td>\n",
       "      <td>8839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>breast</td>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "      <td>353</td>\n",
       "      <td>199659</td>\n",
       "      <td>7282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>uterus</td>\n",
       "      <td>113</td>\n",
       "      <td>113</td>\n",
       "      <td>113</td>\n",
       "      <td>71988</td>\n",
       "      <td>2318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>testis</td>\n",
       "      <td>645</td>\n",
       "      <td>645</td>\n",
       "      <td>659</td>\n",
       "      <td>362820</td>\n",
       "      <td>13467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>esophagus_mucosa</td>\n",
       "      <td>588</td>\n",
       "      <td>588</td>\n",
       "      <td>634</td>\n",
       "      <td>328825</td>\n",
       "      <td>12853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>esophagus_muscularis</td>\n",
       "      <td>647</td>\n",
       "      <td>647</td>\n",
       "      <td>688</td>\n",
       "      <td>352030</td>\n",
       "      <td>14065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>brain_amygdala</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>97</td>\n",
       "      <td>56710</td>\n",
       "      <td>1960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>artery_aorta</td>\n",
       "      <td>449</td>\n",
       "      <td>449</td>\n",
       "      <td>471</td>\n",
       "      <td>260013</td>\n",
       "      <td>9701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>brain_cerebellum</td>\n",
       "      <td>521</td>\n",
       "      <td>521</td>\n",
       "      <td>523</td>\n",
       "      <td>288983</td>\n",
       "      <td>10982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>colon_sigmoid</td>\n",
       "      <td>325</td>\n",
       "      <td>325</td>\n",
       "      <td>338</td>\n",
       "      <td>183948</td>\n",
       "      <td>7036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>kidney_cortex</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>20425</td>\n",
       "      <td>640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ovary</td>\n",
       "      <td>155</td>\n",
       "      <td>155</td>\n",
       "      <td>161</td>\n",
       "      <td>104110</td>\n",
       "      <td>3269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>771</td>\n",
       "      <td>771</td>\n",
       "      <td>822</td>\n",
       "      <td>412184</td>\n",
       "      <td>16804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>brain_frontal_cortex</td>\n",
       "      <td>190</td>\n",
       "      <td>190</td>\n",
       "      <td>197</td>\n",
       "      <td>116536</td>\n",
       "      <td>3980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>adipose_visceral</td>\n",
       "      <td>453</td>\n",
       "      <td>453</td>\n",
       "      <td>474</td>\n",
       "      <td>255846</td>\n",
       "      <td>9735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>heart_atrial_appendage</td>\n",
       "      <td>416</td>\n",
       "      <td>416</td>\n",
       "      <td>435</td>\n",
       "      <td>244342</td>\n",
       "      <td>8970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>nerve_tibial</td>\n",
       "      <td>945</td>\n",
       "      <td>945</td>\n",
       "      <td>1012</td>\n",
       "      <td>478890</td>\n",
       "      <td>20967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>brain_cerebellar_hemisphere</td>\n",
       "      <td>377</td>\n",
       "      <td>377</td>\n",
       "      <td>378</td>\n",
       "      <td>220027</td>\n",
       "      <td>7840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>colon_transverse</td>\n",
       "      <td>384</td>\n",
       "      <td>384</td>\n",
       "      <td>403</td>\n",
       "      <td>231886</td>\n",
       "      <td>8181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>lung</td>\n",
       "      <td>499</td>\n",
       "      <td>499</td>\n",
       "      <td>526</td>\n",
       "      <td>272394</td>\n",
       "      <td>10855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>artery_coronary</td>\n",
       "      <td>167</td>\n",
       "      <td>167</td>\n",
       "      <td>170</td>\n",
       "      <td>105181</td>\n",
       "      <td>3480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>brain_cortex</td>\n",
       "      <td>264</td>\n",
       "      <td>264</td>\n",
       "      <td>273</td>\n",
       "      <td>154969</td>\n",
       "      <td>5543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>esophagus_gej</td>\n",
       "      <td>398</td>\n",
       "      <td>398</td>\n",
       "      <td>410</td>\n",
       "      <td>229489</td>\n",
       "      <td>8404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>brain_anterior_cingulate_cortex</td>\n",
       "      <td>134</td>\n",
       "      <td>134</td>\n",
       "      <td>137</td>\n",
       "      <td>86688</td>\n",
       "      <td>2760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>thyroid</td>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>1000</td>\n",
       "      <td>459133</td>\n",
       "      <td>20582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>skin_sun_exposed</td>\n",
       "      <td>682</td>\n",
       "      <td>682</td>\n",
       "      <td>742</td>\n",
       "      <td>367595</td>\n",
       "      <td>15098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>brain_substantia_nigra</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>85</td>\n",
       "      <td>57097</td>\n",
       "      <td>1760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>blood</td>\n",
       "      <td>629</td>\n",
       "      <td>629</td>\n",
       "      <td>657</td>\n",
       "      <td>339959</td>\n",
       "      <td>13701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>small_intestine</td>\n",
       "      <td>167</td>\n",
       "      <td>167</td>\n",
       "      <td>173</td>\n",
       "      <td>107508</td>\n",
       "      <td>3540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>prostate</td>\n",
       "      <td>196</td>\n",
       "      <td>196</td>\n",
       "      <td>196</td>\n",
       "      <td>122088</td>\n",
       "      <td>4020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>skin_not_sun_exposed</td>\n",
       "      <td>547</td>\n",
       "      <td>547</td>\n",
       "      <td>580</td>\n",
       "      <td>313403</td>\n",
       "      <td>11783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>brain_putamen</td>\n",
       "      <td>202</td>\n",
       "      <td>202</td>\n",
       "      <td>208</td>\n",
       "      <td>125088</td>\n",
       "      <td>4240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>brain_spinal_cord</td>\n",
       "      <td>142</td>\n",
       "      <td>142</td>\n",
       "      <td>142</td>\n",
       "      <td>100582</td>\n",
       "      <td>2940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>adrenal_gland</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>168876</td>\n",
       "      <td>5957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>minor_salivary_gland</td>\n",
       "      <td>137</td>\n",
       "      <td>137</td>\n",
       "      <td>142</td>\n",
       "      <td>99188</td>\n",
       "      <td>2880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>spleen</td>\n",
       "      <td>394</td>\n",
       "      <td>394</td>\n",
       "      <td>407</td>\n",
       "      <td>234160</td>\n",
       "      <td>8313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>liver</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>163</td>\n",
       "      <td>97222</td>\n",
       "      <td>3360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>LCL</td>\n",
       "      <td>148</td>\n",
       "      <td>148</td>\n",
       "      <td>150</td>\n",
       "      <td>100831</td>\n",
       "      <td>3060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>brain_hippocampus</td>\n",
       "      <td>113</td>\n",
       "      <td>113</td>\n",
       "      <td>114</td>\n",
       "      <td>70914</td>\n",
       "      <td>2320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>muscle</td>\n",
       "      <td>597</td>\n",
       "      <td>597</td>\n",
       "      <td>645</td>\n",
       "      <td>320639</td>\n",
       "      <td>13174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           celltype  pos_genes  pos_genes_with_neg_ct  \\\n",
       "0                     brain_caudate        237                    237   \n",
       "1              adipose_subcutaneous        695                    695   \n",
       "2              heart_left_ventricle        364                    364   \n",
       "3                     artery_tibial        684                    684   \n",
       "4                            vagina         83                     83   \n",
       "5                brain_hypothalamus        134                    134   \n",
       "6           brain_nucleus_accumbens        203                    203   \n",
       "7                           stomach        299                    299   \n",
       "8                         pituitary        260                    260   \n",
       "9                          pancreas        421                    421   \n",
       "10                           breast        334                    334   \n",
       "11                           uterus        113                    113   \n",
       "12                           testis        645                    645   \n",
       "13                 esophagus_mucosa        588                    588   \n",
       "14             esophagus_muscularis        647                    647   \n",
       "15                   brain_amygdala         93                     93   \n",
       "16                     artery_aorta        449                    449   \n",
       "17                 brain_cerebellum        521                    521   \n",
       "18                    colon_sigmoid        325                    325   \n",
       "19                    kidney_cortex         32                     32   \n",
       "20                            ovary        155                    155   \n",
       "21                       fibroblast        771                    771   \n",
       "22             brain_frontal_cortex        190                    190   \n",
       "23                 adipose_visceral        453                    453   \n",
       "24           heart_atrial_appendage        416                    416   \n",
       "25                     nerve_tibial        945                    945   \n",
       "26      brain_cerebellar_hemisphere        377                    377   \n",
       "27                 colon_transverse        384                    384   \n",
       "28                             lung        499                    499   \n",
       "29                  artery_coronary        167                    167   \n",
       "30                     brain_cortex        264                    264   \n",
       "31                    esophagus_gej        398                    398   \n",
       "32  brain_anterior_cingulate_cortex        134                    134   \n",
       "33                          thyroid        913                    913   \n",
       "34                 skin_sun_exposed        682                    682   \n",
       "35           brain_substantia_nigra         83                     83   \n",
       "36                            blood        629                    629   \n",
       "37                  small_intestine        167                    167   \n",
       "38                         prostate        196                    196   \n",
       "39             skin_not_sun_exposed        547                    547   \n",
       "40                    brain_putamen        202                    202   \n",
       "41                brain_spinal_cord        142                    142   \n",
       "42                    adrenal_gland        286                    286   \n",
       "43             minor_salivary_gland        137                    137   \n",
       "44                           spleen        394                    394   \n",
       "45                            liver        162                    162   \n",
       "46                              LCL        148                    148   \n",
       "47                brain_hippocampus        113                    113   \n",
       "48                           muscle        597                    597   \n",
       "\n",
       "    total_pos  total_matched_neg  reduced_matched_neg  \n",
       "0         245             145892                 4926  \n",
       "1         748             381142                15206  \n",
       "2         364             225530                 7620  \n",
       "3         729             371392                14887  \n",
       "4          84              51042                 1686  \n",
       "5         139              87201                 2802  \n",
       "6         207             127664                 4211  \n",
       "7         310             171579                 6342  \n",
       "8         264             161558                 5374  \n",
       "9         435             264308                 8839  \n",
       "10        353             199659                 7282  \n",
       "11        113              71988                 2318  \n",
       "12        659             362820                13467  \n",
       "13        634             328825                12853  \n",
       "14        688             352030                14065  \n",
       "15         97              56710                 1960  \n",
       "16        471             260013                 9701  \n",
       "17        523             288983                10982  \n",
       "18        338             183948                 7036  \n",
       "19         32              20425                  640  \n",
       "20        161             104110                 3269  \n",
       "21        822             412184                16804  \n",
       "22        197             116536                 3980  \n",
       "23        474             255846                 9735  \n",
       "24        435             244342                 8970  \n",
       "25       1012             478890                20967  \n",
       "26        378             220027                 7840  \n",
       "27        403             231886                 8181  \n",
       "28        526             272394                10855  \n",
       "29        170             105181                 3480  \n",
       "30        273             154969                 5543  \n",
       "31        410             229489                 8404  \n",
       "32        137              86688                 2760  \n",
       "33       1000             459133                20582  \n",
       "34        742             367595                15098  \n",
       "35         85              57097                 1760  \n",
       "36        657             339959                13701  \n",
       "37        173             107508                 3540  \n",
       "38        196             122088                 4020  \n",
       "39        580             313403                11783  \n",
       "40        208             125088                 4240  \n",
       "41        142             100582                 2940  \n",
       "42        286             168876                 5957  \n",
       "43        142              99188                 2880  \n",
       "44        407             234160                 8313  \n",
       "45        163              97222                 3360  \n",
       "46        150             100831                 3060  \n",
       "47        114              70914                 2320  \n",
       "48        645             320639                13174  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows = []\n",
    "for celltype in selected_vars:\n",
    "    rows.append({'celltype':celltype, \n",
    "                 'pos_genes':len(set(susie_df.query('celltype == @celltype')['gene'])),\n",
    "                 'pos_genes_with_neg_ct':len(set(negvars.query('celltype == @celltype')['gene']) & set(susie_df.query('celltype == @celltype')['gene'])),\n",
    "                 'total_pos':len(set(susie_df.query('celltype == @celltype')['variant'])),\n",
    "                 'total_matched_neg':len(set(negvars.query('celltype == @celltype').merge(susie_df.query('celltype == @celltype')['gene_id'],on='gene_id')['variant'])),\n",
    "                 'reduced_matched_neg':len(selected_vars[celltype]),\n",
    "                 })\n",
    "    \n",
    "pd.DataFrame(rows)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3544e2a-c7a2-4723-b926-8063c3d7e293",
   "metadata": {},
   "source": [
    "## Combine positive and negative variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "383f62a7-4df4-485d-ac2e-245c34eacc90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "542728"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "susie_df = pd.concat([susie_df, negvars_reduced])\n",
    "len(susie_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d3c3cef4-29fc-4519-806d-4bcde89e54e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "susie_df.drop(columns=['molecular_trait_id', 'maf'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "61c3afd7-5d21-4b84-9589-0b5fc52954c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>variant</th>\n",
       "      <th>rsid</th>\n",
       "      <th>cs_size</th>\n",
       "      <th>pip</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>...</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene</th>\n",
       "      <th>dataset</th>\n",
       "      <th>celltype</th>\n",
       "      <th>gene_window_start</th>\n",
       "      <th>gene_window_end</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_mask_start</th>\n",
       "      <th>pos_relative</th>\n",
       "      <th>abspos_rel_TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000079335</td>\n",
       "      <td>ENSG00000079335_L1</td>\n",
       "      <td>chr1_100353172_T_G</td>\n",
       "      <td>rs17420882</td>\n",
       "      <td>4</td>\n",
       "      <td>0.928287</td>\n",
       "      <td>1.472530e-07</td>\n",
       "      <td>-0.261039</td>\n",
       "      <td>0.049038</td>\n",
       "      <td>-5.419992</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>CDC14A</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>100181161.0</td>\n",
       "      <td>100705449.0</td>\n",
       "      <td>+</td>\n",
       "      <td>163840.0</td>\n",
       "      <td>172010</td>\n",
       "      <td>8170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000162631</td>\n",
       "      <td>ENSG00000162631_L1</td>\n",
       "      <td>chr1_107135646_G_C</td>\n",
       "      <td>rs115668827</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.960280e-41</td>\n",
       "      <td>1.297660</td>\n",
       "      <td>0.088815</td>\n",
       "      <td>14.796299</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>NTNG1</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>106976167.0</td>\n",
       "      <td>107500455.0</td>\n",
       "      <td>+</td>\n",
       "      <td>163840.0</td>\n",
       "      <td>159478</td>\n",
       "      <td>4362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000181754</td>\n",
       "      <td>ENSG00000181754_L1</td>\n",
       "      <td>chr1_109509517_A_G</td>\n",
       "      <td>rs2570972</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.708010e-60</td>\n",
       "      <td>0.710806</td>\n",
       "      <td>0.038266</td>\n",
       "      <td>18.768437</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>AMIGO1</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>109149290.0</td>\n",
       "      <td>109673578.0</td>\n",
       "      <td>-</td>\n",
       "      <td>163840.0</td>\n",
       "      <td>164061</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000134184</td>\n",
       "      <td>ENSG00000134184_L3</td>\n",
       "      <td>chr1_109671748_C_T</td>\n",
       "      <td>rs72705222</td>\n",
       "      <td>1</td>\n",
       "      <td>0.998111</td>\n",
       "      <td>3.595720e-06</td>\n",
       "      <td>0.523832</td>\n",
       "      <td>0.111933</td>\n",
       "      <td>4.730494</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>109523974.0</td>\n",
       "      <td>110048262.0</td>\n",
       "      <td>+</td>\n",
       "      <td>163840.0</td>\n",
       "      <td>147773</td>\n",
       "      <td>16067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000134184</td>\n",
       "      <td>ENSG00000134184_L1</td>\n",
       "      <td>chr1_109675302_G_A</td>\n",
       "      <td>rs611951</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.976550e-23</td>\n",
       "      <td>-0.803308</td>\n",
       "      <td>0.077334</td>\n",
       "      <td>-10.478578</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>QTD000116</td>\n",
       "      <td>adipose_subcutaneous</td>\n",
       "      <td>109523974.0</td>\n",
       "      <td>110048262.0</td>\n",
       "      <td>+</td>\n",
       "      <td>163840.0</td>\n",
       "      <td>151327</td>\n",
       "      <td>12513</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id               cs_id             variant         rsid  \\\n",
       "0  ENSG00000079335  ENSG00000079335_L1  chr1_100353172_T_G   rs17420882   \n",
       "1  ENSG00000162631  ENSG00000162631_L1  chr1_107135646_G_C  rs115668827   \n",
       "2  ENSG00000181754  ENSG00000181754_L1  chr1_109509517_A_G    rs2570972   \n",
       "3  ENSG00000134184  ENSG00000134184_L3  chr1_109671748_C_T   rs72705222   \n",
       "4  ENSG00000134184  ENSG00000134184_L1  chr1_109675302_G_A     rs611951   \n",
       "\n",
       "   cs_size       pip        pvalue      beta        se          z  ...  alt  \\\n",
       "0        4  0.928287  1.472530e-07 -0.261039  0.049038  -5.419992  ...    G   \n",
       "1        1  1.000000  2.960280e-41  1.297660  0.088815  14.796299  ...    C   \n",
       "2        1  1.000000  1.708010e-60  0.710806  0.038266  18.768437  ...    G   \n",
       "3        1  0.998111  3.595720e-06  0.523832  0.111933   4.730494  ...    T   \n",
       "4        1  1.000000  2.976550e-23 -0.803308  0.077334 -10.478578  ...    A   \n",
       "\n",
       "     gene    dataset              celltype gene_window_start gene_window_end  \\\n",
       "0  CDC14A  QTD000116  adipose_subcutaneous       100181161.0     100705449.0   \n",
       "1   NTNG1  QTD000116  adipose_subcutaneous       106976167.0     107500455.0   \n",
       "2  AMIGO1  QTD000116  adipose_subcutaneous       109149290.0     109673578.0   \n",
       "3   GSTM1  QTD000116  adipose_subcutaneous       109523974.0     110048262.0   \n",
       "4   GSTM1  QTD000116  adipose_subcutaneous       109523974.0     110048262.0   \n",
       "\n",
       "  gene_strand gene_mask_start pos_relative  abspos_rel_TSS  \n",
       "0           +        163840.0       172010            8170  \n",
       "1           +        163840.0       159478            4362  \n",
       "2           -        163840.0       164061             221  \n",
       "3           +        163840.0       147773           16067  \n",
       "4           +        163840.0       151327           12513  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "susie_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de0092f7-4dd0-46b9-a7a2-57e6771aacc6",
   "metadata": {},
   "source": [
    "## Create VEP input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4e8ade08-7db2-4e52-926e-fbe84c19c2a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "229828"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df = susie_df[['chrom', 'pos', 'ref', 'alt', 'variant', 'rsid', 'gene_id', 'gene', 'gene_strand', 'pos_relative']].drop_duplicates(subset=['gene_id','variant'])\n",
    "len(variant_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c238c742-3b81-40c8-99c2-f5bd7f6ccb6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>variant</th>\n",
       "      <th>rsid</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>pos_relative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100353172</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1_100353172_T_G</td>\n",
       "      <td>rs17420882</td>\n",
       "      <td>ENSG00000079335</td>\n",
       "      <td>CDC14A</td>\n",
       "      <td>+</td>\n",
       "      <td>172010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>107135646</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_107135646_G_C</td>\n",
       "      <td>rs115668827</td>\n",
       "      <td>ENSG00000162631</td>\n",
       "      <td>NTNG1</td>\n",
       "      <td>+</td>\n",
       "      <td>159478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109509517</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1_109509517_A_G</td>\n",
       "      <td>rs2570972</td>\n",
       "      <td>ENSG00000181754</td>\n",
       "      <td>AMIGO1</td>\n",
       "      <td>-</td>\n",
       "      <td>164061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109671748</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_109671748_C_T</td>\n",
       "      <td>rs72705222</td>\n",
       "      <td>ENSG00000134184</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>+</td>\n",
       "      <td>147773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109675302</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_109675302_G_A</td>\n",
       "      <td>rs611951</td>\n",
       "      <td>ENSG00000134184</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>+</td>\n",
       "      <td>151327</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom        pos ref alt             variant         rsid          gene_id  \\\n",
       "0  chr1  100353172   T   G  chr1_100353172_T_G   rs17420882  ENSG00000079335   \n",
       "1  chr1  107135646   G   C  chr1_107135646_G_C  rs115668827  ENSG00000162631   \n",
       "2  chr1  109509517   A   G  chr1_109509517_A_G    rs2570972  ENSG00000181754   \n",
       "3  chr1  109671748   C   T  chr1_109671748_C_T   rs72705222  ENSG00000134184   \n",
       "4  chr1  109675302   G   A  chr1_109675302_G_A     rs611951  ENSG00000134184   \n",
       "\n",
       "     gene gene_strand  pos_relative  \n",
       "0  CDC14A           +        172010  \n",
       "1   NTNG1           +        159478  \n",
       "2  AMIGO1           -        164061  \n",
       "3   GSTM1           +        147773  \n",
       "4   GSTM1           +        151327  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69cb99a5-1be7-4685-9328-682d27cac6bc",
   "metadata": {},
   "source": [
    "## Save final variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b4ccca93-481f-40ae-9f08-16356cc3a397",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df.to_csv(\"variants_df.csv\", index=False)\n",
    "susie_df.to_csv(\"susie_df.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8e4e4e20-0219-49bf-a8b2-6e6daf9de843",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "229828"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(variant_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
