{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "94b0667f-1282-4f43-ba99-7984fef55c62",
   "metadata": {},
   "source": [
    "# Get the pesudobulk data from scooby"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a1cf254-512e-4f6e-846c-e6e8e378f7e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/opt/conda/lib/python3.11/site-packages/anndata/utils.py:434: FutureWarning: Importing read_elem from `anndata.experimental` is deprecated. Import anndata.io.read_elem instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/opt/conda/lib/python3.11/site-packages/anndata/utils.py:434: FutureWarning: Importing sparse_dataset from `anndata.experimental` is deprecated. Import anndata.io.sparse_dataset instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "import polars as pl\n",
    "from torch.utils.data import DataLoader\n",
    "from enformer_pytorch.data import GenomeIntervalDataset\n",
    "from scooby.data import onTheFlyPseudobulkDataset\n",
    "from scooby.utils.utils import undo_squashed_scale, get_gene_slice_and_strand\n",
    "from scooby.utils.transcriptome import Transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b9164b8e-aff1-4b69-9d5a-7bbf6fbf64dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/gstore/data/resbioai/grelu/decima/scooby/'  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1324b3f8-3579-42fa-8a8b-53230d00219b",
   "metadata": {},
   "source": [
    "## Read cell types used for training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9b85c161-8c66-449a-8760-4d2a847bb10d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>cellindex</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B1_B</td>\n",
       "      <td>[5, 9, 20, 32, 112, 128, 151, 265, 294, 360, 3...</td>\n",
       "      <td>1747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD14+_Mono</td>\n",
       "      <td>[1, 11, 13, 19, 30, 38, 49, 50, 51, 58, 62, 64...</td>\n",
       "      <td>10338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CD16+_Mono</td>\n",
       "      <td>[4, 17, 94, 315, 329, 370, 698, 709, 928, 936,...</td>\n",
       "      <td>1762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CD4+_T_activated</td>\n",
       "      <td>[8, 24, 28, 40, 45, 48, 55, 63, 68, 75, 76, 82...</td>\n",
       "      <td>5157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CD4+_T_naive</td>\n",
       "      <td>[7, 44, 47, 54, 56, 59, 88, 116, 123, 132, 140...</td>\n",
       "      <td>4170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           celltype                                          cellindex   size\n",
       "4              B1_B  [5, 9, 20, 32, 112, 128, 151, 265, 294, 360, 3...   1747\n",
       "1        CD14+_Mono  [1, 11, 13, 19, 30, 38, 49, 50, 51, 58, 62, 64...  10338\n",
       "3        CD16+_Mono  [4, 17, 94, 315, 329, 370, 698, 709, 928, 936,...   1762\n",
       "7  CD4+_T_activated  [8, 24, 28, 40, 45, 48, 55, 63, 68, 75, 76, 82...   5157\n",
       "6      CD4+_T_naive  [7, 44, 47, 54, 56, 59, 88, 116, 123, 132, 140...   4170"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_type_index = pd.read_parquet(os.path.join(data_path, 'training_data', 'scooby_training_data', 'celltype_fixed.pq'))\n",
    "cell_type_index['size'] = cell_type_index['cellindex'].apply(lambda x: len(x))\n",
    "cell_type_index['celltype'] = cell_type_index['celltype'].str.replace(' ', '_').replace(r\"G/M_prog\", \"G+M_prog\").replace(\"MK/E_prog\", \"MK+E_prog\") #+ '_filtered'\n",
    "cell_type_index = cell_type_index.sort_values('celltype')\n",
    "cell_type_index.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2a79de2-7f55-4861-a416-6753c3e69c4d",
   "metadata": {},
   "source": [
    "## Load the genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "263ec917-ca00-4f8c-b8bf-5d376b521998",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = os.path.join(data_path, \"gencode.v32.annotation.sorted.gtf.gz\")\n",
    "transcriptome = Transcriptome(gtf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f457e10d-712d-452e-87f7-623c5d8aa781",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = os.path.join(data_path, 'training_data', \"scooby_training_data\", \"genome_human.fa\")\n",
    "base_path = os.path.join(data_path,'training_data', 'scooby_training_data', 'pseudobulks')\n",
    "genes_file = os.path.join(data_path,'training_data','scooby_training_data', 'train_val_test_gene_sequences.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "51489b07-261e-48f9-a450-0125fce31cfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25275\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>663145</td>\n",
       "      <td>859753</td>\n",
       "      <td>AL669831.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>696101</td>\n",
       "      <td>892709</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>720299</td>\n",
       "      <td>916907</td>\n",
       "      <td>FAM87B</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>748365</td>\n",
       "      <td>944973</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>812100</td>\n",
       "      <td>1008708</td>\n",
       "      <td>AL645608.6</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1        2           3  4\n",
       "0  chr1  663145   859753  AL669831.2  +\n",
       "1  chr1  696101   892709   LINC01409  +\n",
       "2  chr1  720299   916907      FAM87B  +\n",
       "3  chr1  748365   944973   LINC01128  +\n",
       "4  chr1  812100  1008708  AL645608.6  +"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_coords = pd.read_table(genes_file, header=None)\n",
    "print(len(gene_coords))\n",
    "gene_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec032bae-da7f-43f5-b214-8b09f2b87c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import anndata\n",
    "#genes = anndata.read_h5ad('/gstore/data/resbioai/grelu/decima/)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81813cba-951a-4e04-9c47-6248eb584e60",
   "metadata": {},
   "source": [
    "## Get the pseudobulk labels used for training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "211f9fb3-e3bc-4323-a3d6-97229dadfef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = GenomeIntervalDataset(\n",
    "    bed_file = genes_file,\n",
    "    fasta_file = fasta_file,\n",
    "    filter_df_fn = lambda df: df.filter((pl.col('column_2') >=0)), \n",
    "    shift_augs = (0,0),\n",
    "    return_augs = True,\n",
    "    context_length = 524288,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "25e89aee-2cad-49a8-ba10-9c2fda850b3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_targets = onTheFlyPseudobulkDataset(\n",
    "    cell_types = cell_type_index['celltype'].values,\n",
    "    ds = ds, \n",
    "    base_path = base_path,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f3815e7f-651d-40bb-8e2b-da863d933e1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25263\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 30%|████████████████████                                              | 7661/25263 [10:05<24:17, 12.08it/s]IOPub message rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_msg_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_msg_rate_limit=1000.0 (msgs/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dataset_target_loader = iter(DataLoader(dataset_targets, batch_size=1, shuffle = False))\n",
    "print(len(dataset_target_loader))\n",
    "\n",
    "all_targets = []\n",
    "gene_names = []\n",
    "\n",
    "for i in tqdm(range(len(dataset_target_loader))):\n",
    "    gene_slice, strand = get_gene_slice_and_strand(transcriptome, dataset_targets.genome_ds.df[i, 'column_4'], dataset_targets.genome_ds.df[i, 'column_2'], span = False)\n",
    "    targets = (next(dataset_target_loader)[2]).float().cuda()\n",
    "    if len(gene_slice) == 0:\n",
    "        continue\n",
    "\n",
    "    if strand == '+':\n",
    "        t = targets[0, gene_slice, ::2]\n",
    "    elif strand == '-':\n",
    "        t = targets[0, gene_slice, 1::2]\n",
    "\n",
    "    all_targets.append(undo_squashed_scale(t, clip_soft=384).sum(axis=0).detach().clone().cpu().squeeze())\n",
    "    gene_names.append(val_dataset_targets.genome_ds.df[i, 'column_4'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "eb862207-55f6-4bf2-9e17-5a37ec50895b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(25187, 21)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_targets = torch.vstack(all_targets).clone().numpy(force=True)\n",
    "all_targets.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "db7843bf-9c5e-448d-85dc-1b4a95ab5a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "torch.save(all_targets, \"count_target_test_no_neighbor.pq\")\n",
    "pd.DataFrame(gene_names).to_parquet(\"gene_names.pq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74d56904-ce66-4038-91b6-47e88b94cf90",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
